File: 0557-reverse-words-in-a-string-iii/0557-reverse-words-in-a-string-iii.java
import java.util.Stack;

class Solution {
    public String reverseWords(String s) {
        int n = s.length();
        StringBuilder result = new StringBuilder();
        Stack<Character> stk = new Stack<>();
        for (int i = 0, j = 0; i < n; i++) 
        {
            if (s.charAt(i)==' ') 
            {
                while (!stk.isEmpty()) 
                {
                    result.append(stk.pop());
                    j++;
                }
                result.append(" ");
            } 
            else 
            {
                stk.push(s.charAt(i));
            }
        }
        while (!stk.isEmpty()) 
        {
            result.append(stk.pop());
        }
        return result.toString();
    }
}

File: 0557-reverse-words-in-a-string-iii/0557-reverse-words-in-a-string-iii.py
class Solution:
    def reverseWords(self, s: str) -> str:
        n, start, end, i = len(s), 0, 0, 0
        result = ""
        while i < n:
            if s[i] == ' ':
                for j in range(end - 1, start - 1, -1):
                    result += s[j]
                result += " "
                start = i + 1
            i += 1
            end = i
        for j in range(end - 1, start - 1, -1):
            result += s[j]
        return result